src/octree/batch_iterator.rs

```diff
@@ -0,0 +1,206 @@
+use collision::Aabb3;
+use fnv::FnvHashMap;
+
+/// size for batch
+pub const NUM_POINTS_PER_BATCH: usize = 500_000;
+
+///possible kind of iterators that can be evaluated in batch of points in BatchIterator
+#[allow(clippy::large_enum_variant)]
+pub enum PointLocation {
+    Any(),
+    Aabb(Aabb3<f64>),
+    Frustum(Matrix4<f64>),
+    OrientedBeam(OrientedBeam),
+}
+
+/// current implementation of the stream of points used in BatchIterator
+struct PointStream<'a, F>
+where
+    F: FnMut(PointData) -> Result<()>,
+{
+    position: Vec<Vector3<f64>>,
+    color: Vec<Vector4<u8>>,
+    intensity: Vec<f32>,
+    func: &'a mut F,
+}
+
+impl<'a, F> PointStream<'a, F>
+where
+    F: FnMut(PointData) -> Result<()>,
+{
+    fn new(num_points_per_batch: usize, func: &'a mut F) -> Self {
+        PointStream {
+            position: Vec::with_capacity(num_points_per_batch),
+            color: Vec::with_capacity(num_points_per_batch),
+            intensity: Vec::with_capacity(num_points_per_batch),
+            func,
+        }
+    }
+
+    /// push point in batch
+    fn push_point(&mut self, point: Point) {
+        self.position.push(point.position);
+        self.color.push(Vector4::new(
+            point.color.red,
+            point.color.green,
+            point.color.blue,
+            point.color.alpha,
+        ));
+        if let Some(point_intensity) = point.intensity {
+            self.intensity.push(point_intensity);
+        };
+    }
+
+    /// execute function on batch of points
+    fn callback(&mut self) -> Result<()> {
+        if self.position.is_empty() {
+            return Ok(());
+        }
+
+        let mut layers = FnvHashMap::default();
+        layers.insert(
+            "color".to_string(),
+            LayerData::U8Vec4(self.color.split_off(0)),
+        );
+        if !self.intensity.is_empty() {
+            layers.insert(
+                "intensity".to_string(),
+                LayerData::F32(self.intensity.split_off(0)),
+            );
+        }
+        let point_data = PointData {
+            position: self.position.split_off(0),
+            layers,
+        };
+        (self.func)(point_data)
+    }
+}
+
+/// Iterator on point batches
+pub struct BatchIterator<'a> {
+    _octree: &'a Octree,
+    iterator: Box<Iterator<Item = Point> + 'a>,
+    batch_size: usize,
+}
+
+impl<'a> BatchIterator<'a> {
+    pub fn new(
+        octree: &'a octree::Octree,
+        location: &'a octree::batch_iterator::PointLocation,
+        size: usize,
+    ) -> Self {
+        let iter: Box<Iterator<Item = Point>> = match location {
+            PointLocation::Any() => Box::new(octree.all_points()),
+            PointLocation::Aabb(aabb) => Box::new(octree.points_in_box(aabb)),
+            PointLocation::Frustum(frustum) => Box::new(octree.points_in_frustum(frustum)),
+            PointLocation::OrientedBeam(beam) => Box::new(octree.points_in_oriented_beam(beam)),
+        };
+        BatchIterator {
```

Comment:
-> `BatchIterator { octree, iterator, batch_size }`

Reply:
done!

Reply:
..sort of `BatchIterator { _octree: octree, iterator, batch_size }`

Reply:
done